src/coreclr/vm/methodtable.cpp

```diff
@@ -4220,19 +4221,31 @@ OBJECTREF MethodTable::GetManagedClassObject()
         // Make sure that we have been restored
         CheckRestore();
 
-        REFLECTCLASSBASEREF  refClass = NULL;
+        REFLECTCLASSBASEREF refClass = NULL;
         GCPROTECT_BEGIN(refClass);
-        refClass = (REFLECTCLASSBASEREF) AllocateObject(g_pRuntimeTypeClass);
 
-        LoaderAllocator *pLoaderAllocator = GetLoaderAllocator();
-
-        ((ReflectClassBaseObject*)OBJECTREFToObject(refClass))->SetType(TypeHandle(this));
-        ((ReflectClassBaseObject*)OBJECTREFToObject(refClass))->SetKeepAlive(pLoaderAllocator->GetExposedObject());
+        LOADERHANDLE exposedClassObjectHandle = NULL;
+        LoaderAllocator* pLoaderAllocator = GetLoaderAllocator();
+        if (!pLoaderAllocator->CanUnload())
+        {
+            // Allocate RuntimeType on a frozen segment
+            FrozenObjectHeapManager* foh = SystemDomain::GetFrozenObjectHeapManager();
+            size_t objSize = g_pRuntimeTypeClass->GetBaseSize();
+            Object* obj = foh->TryAllocateObject(g_pRuntimeTypeClass, objSize);
+            _ASSERTE(obj != NULL);
+            refClass = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(obj);
+            exposedClassObjectHandle = (LOADERHANDLE)obj;
+        }
+        else
+        {
+            refClass = (REFLECTCLASSBASEREF)AllocateObject(g_pRuntimeTypeClass);
+            refClass->SetKeepAlive(pLoaderAllocator->GetExposedObject());
+            exposedClassObjectHandle = pLoaderAllocator->AllocateHandle(refClass);
+        }
+        refClass->SetType(TypeHandle(this));
 
         // Let all threads fight over who wins using InterlockedCompareExchange.
         // Only the winner can set m_ExposedClassObject from NULL.
-        LOADERHANDLE exposedClassObjectHandle = pLoaderAllocator->AllocateHandle(refClass);
-
         if (InterlockedCompareExchangeT(&GetWriteableDataForWrite()->m_hExposedClassObject, exposedClassObjectHandle, static_cast<LOADERHANDLE>(NULL)))
         {
```

Comment:
This backout if not correct.

Do you plan to implement the lock mentioned in the other comment?

Reply:
Ah, yes, it was supposed to be lock, I was just not sure where to define Crst object